Function references + function type (#931)
hhvm/user-documentation: guides/hack/05-functions/30-function-references.md

It is often useful to refer to a function
(or a [method](https://docs.hhvm.com/hack/classes/methods))
without actually calling it&mdash;for example,
as an argument for functions like `Vec\map`.

**Note:** The following syntax is only supported since HHVM 4.79. For older HHVM

> version, see [old syntax](#old-syntax) below.

Reviewer: s/version/versions/

To refer to a top-level (global) function named `foo`, you can write:

```
foo<>; // a reference to global function 'foo'
```

You can think of it like it’s a function call with empty generics, but the list
of arguments has been omitted (the missing parentheses).

The following example stores a function reference in a variable and later calls
the function. Note that the type checker keeps track of the
[function's type](/hack/built-in-types/function)
and correctly checks all calls.

> ```

Reviewer: At least this example should be split out to a separate file and tested

function foo(string $x): bool { ... }

$x = foo<>;     // $x:(function(string): bool)
$y = $x('bar'); // $y:bool
$_ = $x(-1);    // error!
```

This syntax supports namespaced names the same way you would refer to them as
part of a function call, so the following function references are all
equivalent:

```
$fun = \Foo\Bar\Baz\derp<>;
```

```
namespace Foo;
$fun = Bar\Baz\derp<>;
```

```
use namespace Foo\Bar\Baz;
$fun = Baz\derp<>;
```

```
use function Foo\Bar\Baz\derp;
$fun = derp<>;
```

## Static methods

Similarly, you can refer to a static method `bar` on a class `MyClass` by using
the familiar method call syntax, without providing the call arguments. Just
append type arguments (like `<>`) to the function call receiver
(like `MyClass::bar`).

```
MyClass::bar<>; // a reference to static method 'MyClass::bar'
```

- **Private/protected** methods can be referenced using this syntax as long as

>   they are accessible to you in your local scope.

Reviewer: Once you have such a reference to a private/protected method, who can invoke it?

- The arity of the type argument list, if it is non-empty (i.e. not `<>`), must
  match the declaration, just like for an ordinary function call.
- The special wildcard specifier `_` may be provided in place of any or all
  erased (non-reified) generic arguments if you want Hack to infer a type
  automatically based on the type parameter’s constraints. Again this is the
  same as for ordinary function calls.

Example with erased generics:

```
function fizz<Ta as num, Tb>(Ta $a, Tb $b): mixed { ... }

$x = fizz<int, string>; // OK
$x(4, 'hello');
$x(-1, false); // error!

$y = fizz<>;
$y(3.14, new C()); // also OK
$y('yo', derp()); // error!

// OK as well
$z = fizz<int, _>;
$z = fizz<_, string>;
$z = fizz<_, _>;

// these all have errors!
fizz<_>;
fizz<string, _>;
fizz<string, int>;
```

Example with [reified generics](/hack/generics/reified-generics):

```
function buzz<reify T as arraykey>(T $x): mixed { ... }

$w = buzz<int>; // OK
$w(42);
$w("goodbye"); // error!

// these all have errors!
buzz<>;
buzz<_>;
buzz<mixed>;
```

## Introspection

Function references can be cached in APC (the name will be resolved again when

> they are rehydrated) or passed to memoized functions. However, other

Reviewer: rehydrated is an unusual term - maybe 'unserialized', 'retrieved from APC' or similar?

$x(4, 'hello');
$x(-1, false); // error!

$y = fizz<>;
$y(3.14, new C()); // also OK
$y('yo', derp()); // error!

// OK as well
$z = fizz<int, _>;
$z = fizz<_, string>;
$z = fizz<_, _>;

// these all have errors!
fizz<_>;
fizz<string, _>;
fizz<string, int>;
```

Example with [reified generics](/hack/generics/reified-generics):

```
function buzz<reify T as arraykey>(T $x): mixed { ... }

$w = buzz<int>; // OK
$w(42);
$w("goodbye"); // error!

// these all have errors!
buzz<>;
buzz<_>;
buzz<mixed>;
```

## Introspection

Function references can be cached in APC (the name will be resolved again when
they are retrieved) or passed to memoized functions. However, other
serialization formats are not supported.

Internally, function/method references are represented using special data types
that are intended to be opaque. This means they cannot (or should not) be cast
directly to a string or another type, or be accessed in any other way besides
calling them.

It is *sometimes* useful to determine what a function reference is pointing to,
e.g. for use in logging messages. For these *limited* use cases, if you know
enough about the input and output formats you expect, HHVM provides the

> following helpers (but note they are not well supported and may change):

Reviewer 2: We don't need to waffle as much in the documentation. These are the only native APIs that do what they do, and the ones you _should_ use, but internally we have our own that wrap them and we recommend the latter instead. Maybe we can add an FB-only blurb that mentions this?

Author: Sounds reasonable, do you have a suggestion on what exactly to put here? I'm not sure how much we want to encourage/discourage people from using these. If they're unsupported and likely to change, we should probably mention that.

We can add a FB-only block, but we're only hiding those using CSS so shouldn't mention anything sensitive.

> It is *sometimes* useful to determine what a function reference is pointing to,
> e.g. for use in logging messages. For these *limited* use cases, if you know
> enough about the input and output formats you expect, HHVM provides the
> following helpers (but note they are not well supported and may change):

Reviewer 2: Nothing too sensitive here. The `HackCallable` class is what I would recommend for internal use.

```suggestion
If you need to determine what a function reference is pointing to, e.g. for
use in logging messages, and you know enough about the expected input and
output formats, HHVM provides the following helpers (but note they are not
well supported and may change):
```

* `HH\is_fun`
* `HH\fun_get_function`
* `HH\is_class_meth`
* `HH\class_meth_get_class`

> * `HH\class_meth_get_method`

Reviewer 2: Are there pages for these that we can link to? We probably need to add docblocks to their definitions though.

Author: Yep, these are linkified automatically. Link goes to https://docs.hhvm.com/hack/reference/function/HH.is_fun/ etc.

Docblocks could be nice but I think it's pretty clear what each one does in the context of this guide, so not really necessary.

Be very careful and deliberate when using these, as they are loosely typed but
will throw exceptions for bad arguments. Use of reflection APIs in production is

> also discouraged because it often has a large performance impact.

Reviewer 2: Is this still true when considering how most people run HHVM outside of FB?

Author: I think it's true in general that "using reflection in production is slow" but these APIs are no worse than anything else (probably even much better than things like ReflectionFunction...) so you're probably right, it doesn't make a lot of sense to mention it here. I'll remove it.
